src/PMNS_NUNM.cxx

```diff
@@ -0,0 +1,380 @@
+        Alpha(i, j) *= 1 / std::sqrt(X(i, i).real()); // Scale by the inverse square root of the diagonal elements of X
+      }
+    }
+  }
+  
+  // Solve for eigensystem
+  SolveHam();
+  
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+	Evec0(i, j) = fEvec[i][j]; //
+    }
+  }
+  
+  if (isFirst) { EvecA = Evec0.adjoint() * Alpha.adjoint(); }
+  else{ EvecA = Evec0.adjoint();}
+  
+  if (isLast) { Evec = Alpha * Evec0; }
+  else{ Evec = Evec0 ;}
+  
+  double LengthIneV = kKm2eV * p.length;
+  for (int i = 0; i < fNumNus; i++) {
+    double arg = fEval[i] * LengthIneV;
+    fPhases[i] = complexD(cos(arg), -sin(arg));
+  }
+  
+  for (int i = 0; i < fNumNus; i++) {
+    fBuffer[i] = 0;
+    for (int j = 0; j < fNumNus; j++) {
+      fBuffer[i] += EvecA(i,j) * fNuState[j];
+    }
+    fBuffer[i] *= fPhases[i];
+  }
+
+  // Propagate neutrino state
+  for (int i = 0; i < fNumNus; i++) {
+    fNuState[i] = 0;
+    for (int j = 0; j < fNumNus; j++) {
+      fNuState[i] += Evec(i,j) * fBuffer[j];
+    }
+  }
+}
+
+
+//.............................................................................
+///
+/// Build the full Hamiltonian in matter --> from sterile
+///
+/// Here we divide the mass squared matrix Hms by the 2E
+/// to obtain the vacuum Hamiltonian in eV. Then, the matter
+/// potential is added to the electron and NC components.
+///
+/// the neutron matter effect is added to the diag of V
+/// depending on the scenario the writting of the Hamiltonian
+/// in the tilde basis is different
+///
+/// solving the sytem in the tilde basis: H = U·Dm·U~ + Alpha~·V·Alpha
+/// in high scale scenario : H = U·Dm·U~ + Alpha^(-1)·V·Alpha~^(-1)
+/// inspired from https://arxiv.org/pdf/2301.12960.pdf and
+/// tilde basis in eq B5. https://cds.cern.ch/record/1032544/files/PhysRevD.76.093005.pdf
+///
+///
+///
+void PMNS_NUNM::UpdateHam()
+{ 
+  double rho = fPath.density;
+  double zoa = fPath.zoa;
+
+  double lv = 2 * kGeV2eV * fEnergy; // 2*E in eV
+  
+  double kr2GNe = kK2 * M_SQRT2 * kGf * rho * zoa ; // Electron matter potential in eV
+  double kr2GNn = kK2 * M_SQRT2 * kGf * rho * (1. - zoa) / 2. * fracVnc ; // Neutron matter potential in eV
+  
+  // Finish build Hamiltonian in matter with dimension of eV
+  
+  for (int i = 0; i < fNumNus; i++) { 
+      if (!fIsNuBar){ V(i,i) = -1. * kr2GNn;}
+      else { V(i,i) = kr2GNn;}
+      for (int j = 0; j < fNumNus; j++) {
+        if (!fIsNuBar) Ham(i,j) = fHms[i][j] / lv;
+        else Ham(i,j) = conj(fHms[i][j]) / lv;
+      }
+  }
+  
+  if (!fIsNuBar){V(0,0) += kr2GNe;}
+  else {V(0,0) -= kr2GNe;}
+
+  if (fscale == 0){ Ham += Alpha.adjoint()*V*Alpha ;} // low scale scenario with mixing matrix part of larger unitary matrix 
+  else if (fscale == 1){ Ham += Alpha.inverse()*V*(Alpha.adjoint()).inverse();} // high scale scenario with non unitary mixing matrix
+
+  for (int i = 0; i < fNumNus; i++) {
+    for (int j = 0; j < fNumNus; j++) {
+	fHam[i][j] = Ham(i,j);
+    }
+  }
```

Comment:
Why do we need this copy stage?

Reply:
fHam is used in void PMNS_Fast::SolveHam() so in order not to rewrite it I convert Ham to fHam here